fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -815,8 +840,11 @@ def _wait_for_post_backward(self) -> None:
         if self.move_grads_to_cpu:
             # Wait for the non-blocking GPU -> CPU grad transfers to finish.
             torch.cuda.current_stream().synchronize()
-        # A backward pass is done.
-        self.training_state = TrainingState.IDLE
+        # A backward pass is done, update root and nested FSDP's flags.
+        for m in self.modules():  # includes self
```

Comment:
Thanks for the suggestion. Found and fixed a bug on the TrainingState.

Reply:
Nice catch!

Reply:
right now, the delayed reduce test actually catches this. I think there are about 4 tests affected by this. That's how I found it.